api/v1/dois/PKPDoiHandler.inc.php

```diff
@@ -260,11 +260,36 @@ public function edit(SlimRequest $slimRequest, APIResponse $response, array $arg
             return $response->withStatus(400)->withJson($errors);
         }
 
-        Repo::doi()->edit($doi, $params);
+		$pubObjectType = $slimRequest->getParsedBodyParam('pubObjectType');
+		$pubObjectId = $slimRequest->getParsedBodyParam('pubObjectId');
 
-        $doi = Repo::doi()->get($doi->getId());
+		// Default behaviour, only edits DOI
+		if (empty($pubObjectType) && empty($pubObjectId)) {
+			Repo::doi()->edit($doi, $params);
+			$doi = Repo::doi()->get($doi->getId());
 
-        return $response->withJson(Repo::doi()->getSchemaMap()->map($doi), 200);
+			return $response->withJson(Repo::doi()->getSchemaMap()->map($doi), 200);
+		}
+
+		// If there is a pubObjectType and pubObjectId, we want to copy the existing DOI and create a new DOI and
+		// assign it to the pubObject. This prevents data loss when a DOI object is assigned to more than one pubObject.
+
+		// Check pubObject for doiId
+		$pubObject = Repo::doi()->getPubObjectFromType($pubObjectType, $pubObjectId);
+		if ($pubObject->getData('doiId') != $doi->getId()) {
+			return $response->withStatus(404)->withJsonError('api.dois.404.pubObjectNotFound');
+		}
+
+		// Copy DOI object data
+		$newDoi = clone $doi;
+		$newDoi->unsetData('id');
+		$newDoi->setAllData(array_merge($newDoi->getAllData(), $params));
+		$newDoiId = Repo::doi()->add($newDoi);
+
+		Repo::doi()->updatePubObjectDoiFromType($pubObject ,$pubObjectType, $newDoiId);
+		Repo::doi()->removeDanglingObjects($doi->getId());
```

Comment:
Since `getPubObjectFromType()` and `updatePubObjectDoiFromType()` are all part of the funny logic of the API, let's make these protected methods of the handler class. I know you'll need a new `DoiHandler()` for each application, but it should be pretty small. You can also then define some constants for the types there which match the API specifications, without needing to roll this out to the rest of the application.

It might also free you to do a little bit more in the `edit` method and less in the helper methods. I would imagine something like a more general-purpose use function in the repository:

```php
class Repository
{
    public function isAssigned(int $doiId): bool
    {
        $publicationCount = ...;
        $galleyCount = ...;

        return $publicationCount || $galleyCount;
    }
}
```

And then the rest of the logic is in the API handler, and the app-specific API handler only needs to be responsible for returning the right variable for each object type. It doesn't have to take the action:

```php
class PKPDoiHandler extends APIHandler
{
    public const OBJECT_TYPE_SUBMISSION = 'submission';

    protected method getPubObjectRepo(string $type)
    {
        return match($type) {
            self::OBJECT_TYPE_SUBMISSION => Repo::submission(),
        };
    }

    public function edit(...)
    {
        // ...

        $repo = $this->getPubObjectRepo($type);
        if (!$repo) {
            return $response->withStatus(403)->withJsonError('pub.type.not.recognized');
        }
        
        // ...

        $repo->edit(...);

        if (!Repo::doi()->isAssigned($doi->getId()) {
            Repo::doi()->delete($doi);
        }
    }
}
```